os/kernel/binary_manager/binary_manager.c

```diff
@@ -145,6 +154,86 @@ void binary_manager_update_running_state(int bin_id)
 	binary_manager_notify_state_changed(bin_idx, BINARY_STARTED);
 }
 
+#ifdef CONFIG_BINMGR_RECOVERY
+static int faultmsg_sender(int argc, char *argv[])
+{
+	int idx;
+	int ret;
+	faultmsg_t *msg;
+	binmgr_request_t request_msg;
+
+	/* Initialize pre-allocated fault messages */
+
+	for (idx = 0; idx < FAULTMSG_COUNT; idx++) {
+		sq_addlast((FAR sq_entry_t *)&g_prealloc_faultmsg[idx], (FAR sq_queue_t *)&g_freemsg_list);
+	}
+
+	while (1) {
+		/* Wait for fault messages and handle it */
+		up_block_task(this_task(), TSTATE_WAIT_UNBLOCK);
+		while (!sq_empty(&g_faultmsg_list)) {
+			msg = (struct faultmsg_s *)sq_remfirst(&g_faultmsg_list);
+			request_msg.cmd = BINMGR_FAULT;
+			request_msg.requester_pid = msg->binid;
+			bmllvdbg("Send fault message, bin id %d\n", request_msg.requester_pid);
+			ret = mq_send(g_binmgr_mq_fd, (const char *)&request_msg, sizeof(binmgr_request_t), BINMGR_FAULT_PRIO);
+			ASSERT(ret == OK);
+			sq_addlast((FAR sq_entry_t *)msg, (FAR sq_queue_t *)&g_freemsg_list);
+		}
+	}
+
+	return 0;
+}
+
+/* This function is called in fault handler.
+ * Add fault message and Unblock Fault message Sender.
+ */
+int binary_manager_send_faultmsg(pid_t pid)
+{
+	faultmsg_t *msg;
+
+	if (g_faultmsg_sender == NULL) {
+		bmlldbg("Faultmsg sender is not initialized\n");
+		return ERROR;
+	}
+
+	msg = (faultmsg_t *)sq_remfirst((sq_queue_t *)&g_freemsg_list);
+	if (msg == NULL) {
+		bmlldbg("Failed to allocate fault message\n");
+		return ERROR;
+	}
+	msg->binid = pid;
+	sq_addlast((sq_entry_t *)msg, (sq_queue_t *)&g_faultmsg_list);
+	up_unblock_task(g_faultmsg_sender);
+
+	return OK;
+}
+#endif
+
+void binary_manager_add_binlist(FAR struct tcb_s *tcb)
```

Comment:
how about separating these codes? I mean that we had better make a new file and locate all of functions related bin data.